### PR TITLE
Fix unit tests and unmark passing ROCm tests

### PR DESCRIPTION
**Pull request by sudhu2k**
# What does this PR do ?
This PR fixes some unit tests and also unmarks tests which are marked as failing on rocm but are actually passing on rocm.

## Fixes:
Megatron-LM Unit test failure analysis
[#14085](https://github.com/ROCm/frameworks-internal/issues/14085)

## Currently skipped test failure after changes in this PRs:

TestParallelMLAAttention::test_gpu_forward_thd -- Incompatible

TestParallelMLAAttention::test_gpu_forward_thd_precision -- Incompatible

TestContextParallelMLAAttention::test_gpu_forward_thd -- Incompatible

TestContextParallelMLAAttention::test_gpu_forward_thd_precision -- Incompatible

TestContextParallelMLAAttentionPrecision::test_gpu_forward_thd_precision -- Incompatible

TestParallelMLAAttentionPrecisionWithRopeFusion::test_gpu_forward_thd_precision -- Incompatible

tests/unit_tests/dist_checkpointing/test_async_save.py::TestAsyncSave::test_errors_are_reported[write_data_os_err_mock_fn-True] -- Incompatible due to change from fork to spawn

tests/unit_tests/dist_checkpointing/models/test_mlp_glu.py::TestParallelMLPWithGLU::test_oom_is_handled -- Incompatible

### CODE REVIEW

File: Dockerfile_rocm.ci [hunk 1]
```diff
@@ -62,7 +62,7 @@ RUN git clone https://github.com/sudhu2k/mamba mamba &&\
     cd mamba &&\
     git checkout hipcub_warp_threads_deprecation &&\
     git show --oneline -s &&\
-    pip install --no-build-isolation .
+    pip install --no-build-isolation --no-deps .
```
Comment (wenchenvincent) [inline Dockerfile_rocm.ci:65]: Why do we need to add `--no-deps`?

Comment (sudhu2k) [reply Dockerfile_rocm.ci:65]: This package installed triton 3.5.0 as it's dependency which caused several issues. I didn't want to change the triton version while installing mamba.

File: tests/unit_tests/transformer/test_multi_token_prediction.py [hunk 2]
```diff
@@ -42,8 +42,7 @@
 except ImportError:
     HAVE_TE = False
 
-_SEED = 42
```
Comment (wenchenvincent) [inline tests/unit_tests/transformer/test_multi_token_prediction.py:45]: How much is the divergence with the original seed?

Comment (sudhu2k) [reply tests/unit_tests/transformer/test_multi_token_prediction.py:45]: With the original seed.
Mismatched elements: 2/64 (3.1250%)
Max abs diff: 0.02858257293701172
Max rel err: 0.0024164761416614056

--- Worst by absolute difference ---
Index: (0, 26)
ref: 11.828204154968262 out: 11.79962158203125 |diff|: 0.02858257293701172
Tolerance at this element: atol + rtol*|ref| = 0.012828204154968263

--- Worst by relative error ---
Index: (0, 26)
ref: 11.828204154968262 out: 11.79962158203125 |diff|: 0.02858257293701172
rel err: 0.0024164761416614056

Comment (wenchenvincent) [reply tests/unit_tests/transformer/test_multi_token_prediction.py:45]: So the test is using TE grouped gemm by default?

Comment (sudhu2k) [reply tests/unit_tests/transformer/test_multi_token_prediction.py:45]: Yes, that's right.

**Comment by wenchenvincent** [comment]
@sudhu2k Could you make the title of the PR more descriptive?

Also, it seems that the CI was not triggered?

**Comment by sudhu2k** [comment]
Changed the title!
It was getting triggered for code changes/pushes, but since I've been manually re-running the CI a lot, I think the manual runs doesn't notify github of the test results. Might have to check that.
CI:
https://ml-ci-internal.amd.com/job/Megatron-LM/job/Megatron-LM/view/change-requests/job/PR-96/

File: tests/unit_tests/transformer/test_multi_token_prediction.py (hunk 2 above)
Comment (sudhu2k) [reply tests/unit_tests/transformer/test_multi_token_prediction.py:45]: @wenchenvincent 

## Model Weight Shape Inspection (TP=4, CP=2):


Context Parallelism Configuration:
  Expected CP size: 2
  Actual CP size: 2
  CP rank: 0
  ✓ CP size matches expected value

Found 74 TP-sharded weight parameters:
| Parameter Name                                               | Original Shape         | Sharded Shape | Elements |
|--------------------------------------------------------------|------------------------|---------------|----------|
| decoder.layers.0.self_attention.linear_qkv.layer_norm_weight | torch.Size([128])      | (128,)        | 128      |
| decoder.layers.0.self_attention.linear_qkv.weight            | torch.Size([384, 128]) | (96, 128)     | 12288    |
| decoder.layers.0.mlp.experts.linear_fc1.weight0              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc1.weight1              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc1.weight2              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc1.weight3              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc1.weight4              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc1.weight5              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc1.weight6              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc1.weight7              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.0.mlp.experts.linear_fc2.weight0              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.0.mlp.experts.linear_fc2.weight1              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.0.mlp.experts.linear_fc2.weight2              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.0.mlp.experts.linear_fc2.weight3              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.0.mlp.experts.linear_fc2.weight4              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.0.mlp.experts.linear_fc2.weight5              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.0.mlp.experts.linear_fc2.weight6              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.0.mlp.experts.linear_fc2.weight7              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.self_attention.linear_qkv.layer_norm_weight | torch.Size([128])      | (128,)        | 128      |
| decoder.layers.1.self_attention.linear_qkv.weight            | torch.Size([384, 128]) | (96, 128)     | 12288    |
| decoder.layers.1.mlp.experts.linear_fc1.weight0              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc1.weight1              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc1.weight2              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc1.weight3              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc1.weight4              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc1.weight5              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc1.weight6              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc1.weight7              | torch.Size([640, 128]) | (160, 128)    | 20480    |
| decoder.layers.1.mlp.experts.linear_fc2.weight0              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.mlp.experts.linear_fc2.weight1              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.mlp.experts.linear_fc2.weight2              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.mlp.experts.linear_fc2.weight3              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.mlp.experts.linear_fc2.weight4              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.mlp.experts.linear_fc2.weight5              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.mlp.experts.linear_fc2.weight6              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| decoder.layers.1.mlp.experts.linear_fc2.weight7              | torch.Size([128, 320]) | (128, 80)     | 10240    |
| mtp.layers.0.eh_proj.weight                                  | torch.Size([128, 256]) | (32, 256)     | 8192     |
| mtp.layers.0.transformer_layer.self_attention.linear_qkv.layer_norm_weight | torch.Size([128])      | (128,)        | 128      |
| mtp.layers.0.transformer_layer.self_attention.linear_qkv.weight | torch.Size([384, 128]) | (96, 128)     | 12288    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight0 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight1 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight2 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight3 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight4 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight5 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight6 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc1.weight7 | torch.Size([640, 128]) | (160, 128)    | 20480    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc2.weight0 | torch.Size([128, 320]) | (128, 80)     | 10240    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc2.weight1 | torch.Size([128, 320]) | (128, 80)     | 10240    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc2.weight2 | torch.Size([128, 320]) | (128, 80)     | 10240    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc2.weight3 | torch.Size([128, 320]) | (128, 80)     | 10240    |
| mtp.layers.0.transformer_layer.mlp.experts.linear_fc2.weight4 | torch.Size([128, 320]) | (128, 80)     | 10240    |

Context Parallelism Batch Shape Verification:

Batch shapes (CP=2):
Key                  Before CP Split                After CP Split                 Seq Dim Change      
tokens               torch.Size([2, 32])            torch.Size([2, 16])            32 -> 16 ✓          
labels               torch.Size([2, 32])            torch.Size([2, 16])            32 -> 16 ✓          
loss_mask            torch.Size([2, 32])            torch.Size([2, 16])            32 -> 16 ✓          
attention_mask       torch.Size([2, 1, 32, 32])     torch.Size([2, 1, 16, 32])     32 -> 16 ✓          
position_ids         torch.Size([2, 32])            torch.Size([2, 16])            32 -> 16 ✓          
output_ref           torch.Size([2, 32])            torch.Size([2, 16])            32 -> 16 ✓          

I can see that CP and TP are being applied on this test.

**Review by wenchenvincent — APPROVED** [review]
LGTM.
